src/common/random_generator.h

```diff
@@ -0,0 +1,211 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2017 by Contributors
+ * \file random_generator.h
+ * \brief Native random number generator.
+ */
+#ifndef MXNET_COMMON_RANDOM_GENERATOR_H_
+#define MXNET_COMMON_RANDOM_GENERATOR_H_
+
+#include <mxnet/base.h>
+#include <random>
+
+#ifdef __CUDACC__
+#include <curand_kernel.h>
+#endif  // __CUDACC__
+
+using namespace mshadow;
+
+namespace mxnet {
+namespace common {
+namespace random {
+
+// Elementary random number generation for int/uniform/gaussian in CPU and GPU.
+// Will use float data type whenever instantiated for half_t or any other non
+// standard real type.
+template<typename Device, typename DType MSHADOW_DEFAULT_DTYPE>
+class RandGenerator;
+
+template<typename Device, typename DType MSHADOW_DEFAULT_DTYPE>
+class RandGeneratorGlobal;
+
+template<typename xpu, typename DType MSHADOW_DEFAULT_DTYPE>
+void RandGeneratorSeed(Stream<xpu> *, RandGenerator<xpu, DType> *, uint32_t seed);
+
+template<typename xpu, typename DType MSHADOW_DEFAULT_DTYPE>
+RandGenerator<xpu, DType> *NewRandGenerator();
+
+template<typename xpu, typename DType MSHADOW_DEFAULT_DTYPE>
+void DeleteRandGenerator(RandGenerator<xpu, DType> *);
+
+template<typename DType>
+class RandGenerator<cpu, DType> {
+ public:
+  typedef typename std::conditional<std::is_floating_point<DType>::value,
+                                    DType, double>::type FType;
+
+  MSHADOW_XINLINE void Seed(uint32_t seed, uint32_t idx) {
+    engine_.seed(seed);
+  }
+
+  MSHADOW_XINLINE int rand() { return engine_(); }
+
+  MSHADOW_XINLINE FType uniform() {
+    typedef typename std::conditional<std::is_integral<DType>::value,
+                                      std::uniform_int_distribution<DType>,
+                                      std::uniform_real_distribution<FType>>::type GType;
+    GType dist_uniform;
+    return dist_uniform(engine_);
+  }
+
+  MSHADOW_XINLINE FType normal() {
+    std::normal_distribution<FType> dist_normal;
+    return dist_normal(engine_);
+  }
+
+ private:
+  std::mt19937 engine_;
+};
+
+#ifdef __CUDACC__
+// at least how many random numbers should be generated by one GPU thread.
+const int kGPUMinRndNumberPerThread = 64;
+// store how many global random states.
+const int kGPURndStateNum = 32768;
+
+// uniform number generation in Cuda made consistent with stl (include 0 but exclude 1)
+// by using 1.0-curand_uniform(). Needed as some samplers below won't be able to deal with
```

Comment:
Comment taken from prior code. A bit misleading as it references "samplers below" (but in this file, there are none).